nri-prelude/src/Internal/IO.hs

```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE BlockArguments #-}
+
+module Internal.IO
+  ( putText,
+    putTextLn,
+  )
+where
+
+import Control.Concurrent.MVar
+import System.IO.Unsafe (unsafePerformIO)
+import Text
+import qualified Prelude
+
+{-# NOINLINE stdoutLock #-}
+stdoutLock :: MVar ()
```

Comment:
I just want to think through the safety of this.
The [docs for MVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html) say
> the "bigger" functions in this module ([swapMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:swapMVar), [withMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:withMVar), [modifyMVar_](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:modifyMVar_) and [modifyMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:modifyMVar)) are simply the composition of a [takeMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:takeMVar) followed by a [putMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:putMVar) with exception safety. These have atomicity guarantees only if all other threads perform a [takeMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:takeMVar) before a [putMVar](https://hackage.haskell.org/package/base-4.19.1.0/docs/Control-Concurrent-MVar.html#v:putMVar) as well; otherwise, they may block.

We do meet those criteria, so that suggests that we do have atomicity guarantees. And I don't see how it could deadlock.
But I do think that we should have some code comments explaining why we think it's OK. Otherwise someone is going to come across this in a year and wonder.

Also I note that we do not have very many places where we print to stdout. It seems to be just on startup and shutdown of services. We do other things for logging.

Still, it is easy to make mistakes with this stuff, so it does make me feel a bit cautious. I wonder if it's worth having an env var to turn off the locking behaviour.